_posts/function-contracts.md

```diff
@@ -0,0 +1,409 @@
+            // Make sure preconditions are respected for recursion
+            assert!(max != 0 && min != 0);
+            let result = kani::any();
+            kani::assume(max % result == 0 && min % result == 0 && result != 0);
+            result
+        }
+    };
+    // Make sure postconditions hold
+    assert!(max % result == 0 && min % result == 0 && result != 0);
+}
+```
+
+First it is helpful to think about the recursion as a sequence of individual steps that build on top of one another, like a pyramid. Each step performs the same computation, but with different inputs, because the previous step calls into it with `min` and `rest`, which differ from `max` and `min`. Now consider the first step. It is called with non-deterministic inputs that are only constrained by the precondition. Then a computation is performed to calculate `min` and `rest` that is used for the recursive call and at the end the postconditions are enforced with `assert!(max % result == 0 && ...)`. What does that mean? It means if this step passes verification we can be sure that the postconditions hold for the inputs we verified with, e.g. any combination of `max` and `min` that satisfies `max != 0 && min != 0`. Now let's consider again the recursive call. We call with `min` and `rest` which are not the same as `max` and `min`. However remember that actually what our first step verification proves not just that the postconditions hold for *a* `max` and `min`, but in fact for *any* `max` and `min`, so long as the precondition is satisfied. Therefore we can conclude that it will also hold for `gcd(min, rest)`, if `min` and `rest` satisfy the preconditions, e.g. `min != 0 && rest != 0`. And that is precisely what is enforced where the abstraction is employed with `assert!(max != 0 && min != 0)`.
+
+As a small technicality: induction is usually split into the so called "base case" and the "induction step". In our example a single run of the function actually combines both, which is the `if rest == 0` split. When `rest == 0` then we have reached the base case, otherwise we are in the induction step. Both the base case and the induction step must uphold the ponstconditions. The way that our proof is set up here already ensures that because the postcondition `assert!` gets enforced regardless of whether we took the base case or the induction step.[^return] 
+
+[^return]: The astute reader might note that this would not be the case if the body of the function we are verifying contained a `return` statement. Rest assured that the code you see here is jsut an illustration and the actual code generated for contract verification is not so easily fooled. If you want to see actual examples of how this is set up there is one in the implementation documentation of the [contracts macros](https://github.com/model-checking/kani/blob/main/library/kani_macros/src/sysroot/contracts.rs) or you can dump the generated code with the rustc flag `-Zunpretty=expanded`.
```

Comment:
```suggestion
[^return]: The astute reader might note that this would not be the case if the body of the function we are verifying contained a `return` statement. Rest assured that the code you see here is just an illustration and the actual code generated for contract verification is not so easily fooled. If you want to see actual examples of how this is set up there is one in the implementation documentation of the [contracts macros](https://github.com/model-checking/kani/blob/main/library/kani_macros/src/sysroot/contracts.rs) or you can dump the generated code with the rustc flag `-Zunpretty=expanded`.
```